Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE with AudioBufferList

cdumez opened the pull request:
#### cb84985b00a603e34f8688f31305b646df43122f
<pre>
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE with AudioBufferList
<a href="https://bugs.webkit.org/show_bug.cgi?id=303743">https://bugs.webkit.org/show_bug.cgi?id=303743</a>

Reviewed by Darin Adler.

* Source/WTF/wtf/StdLibExtras.h:
(WTF::lazyInitialize):
* Source/WTF/wtf/SystemFree.h:
(WTF::SystemFree::operator() const):
* Source/WTF/wtf/SystemMalloc.h:
(WTF::SystemMallocBase::zeroedMalloc):
* Source/WebCore/Modules/webaudio/MediaStreamAudioSourceCocoa.cpp:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/cf/CoreAudioExtras.h: Renamed from Source/WebCore/platform/audio/cocoa/SpanCoreAudio.h.
(PAL::allocationSize):
(PAL::createAudioBufferList):
* Source/WebCore/PAL/pal/spi/cf/CoreAudioSPI.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.cpp:
* Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.cpp:
(WebCore::AudioFileReader::decodeWebMData const):
(WebCore::AudioFileReader::createBus):
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.mm:
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.cpp:
* Source/WebCore/platform/audio/cocoa/CARingBuffer.cpp:
* Source/WebCore/platform/audio/cocoa/WebAudioBufferList.cpp:
(WebCore::WebAudioBufferList::WebAudioBufferList):
(WebCore::WebAudioBufferList::reset):
* Source/WebCore/platform/audio/cocoa/WebAudioBufferList.h:
* Source/WebCore/platform/audio/mac/AudioHardwareListenerMac.cpp:
* Source/WebCore/platform/audio/mac/AudioSessionMac.mm:
* Source/WebCore/platform/graphics/avfoundation/AudioSourceProviderAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/AudioSourceProviderAVFObjC.mm:
(WebCore::AudioSourceProviderAVFObjC::prepare):
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererInternalUnit.cpp:
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureUnit.cpp:
* Source/WebCore/platform/mediastream/mac/MockAudioCaptureUnit.mm:
* Source/WebCore/platform/mediastream/mac/RealtimeOutgoingAudioSourceCocoa.cpp:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::renderAudio):

Canonical link: <a href="https://commits.webkit.org/304157@main">https://commits.webkit.org/304157@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/67a50f1242dbcb3d880d4f49545c3e8b6df11afc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134730 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7174 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45975 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142250 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86654 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136600 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7784 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7026 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/102968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/70248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137677 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5476 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120755 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83778 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/20866ab2-2cba-4f39-bf1d-3b8652db206c) 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/134079 "Build is in progress. Recent messages:") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5322 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2932 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2839 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126768 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38902 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144945 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/133237 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6850 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39483 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111361 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6921 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5743 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111662 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28324 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5160 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117035 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60745 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6898 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35219 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/166142 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6699 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70479 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43422 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6935 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6808 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/PAL/pal/cf/CoreAudioExtras.h:
```diff
@@ -53,4 +55,27 @@ inline std::span<const AudioBuffer> span(const AudioBufferList& list)
+enum class ShouldZeroMemory : bool { No, Yes };
+inline std::unique_ptr<AudioBufferList, WTF::SystemFree<AudioBufferList>> createAudioBufferList(uint32_t bufferCount, ShouldZeroMemory shouldZeroMemory)
+{
+    CheckedSize bufferListSize = offsetof(AudioBufferList, mBuffers);
+    bufferListSize += CheckedSize { sizeof(AudioBuffer) } * std::max<uint32_t>(1, bufferCount);
```
What’s the goal of the checking here? Crash on overflow?

cdumez replied:
I merely moved this logic from WebAudioBufferList but yes, it looks to me like it is to crash on overflow.

darinadler commented on Source/WebCore/SourcesCocoa.txt:
```diff
@@ -291,7 +291,7 @@ platform/audio/ios/MediaSessionHelperIOS.mm @nonARC @no-unify
 platform/audio/mac/AudioBusMac.mm @nonARC
 platform/audio/mac/AudioOutputUnitAdaptorMac.cpp
-platform/audio/mac/AudioHardwareListenerMac.cpp
+platform/audio/mac/AudioHardwareListenerMac.cpp @no-unify
```
Always annoying to have more of these. Can we avoid this?

cdumez replied:
Hmm, all I did was introduce a new file and then I got an unrelated build error in AudioHardwareListenerMac.cpp due to `kAudioHardwarePropertyProcessIsRunning` being redefined:
```
enum {
    kAudioHardwarePropertyProcessIsRunning = 'prun'
};
```

Several files in the same unified build group must be defining this constant. I'm sure this is avoidable though. Maybe I can move this constant definition to a header somewhere?

cdumez replied:
Actually, I don't think `kAudioHardwarePropertyProcessIsRunning` is defined in another WebKit cpp based on grep. Maybe a system include.

cdumez replied:
I have moved this enum declaration to `CoreAudioSPI.h` for when the internal SDK is not used because it comes from a private header otherwise.

cdumez committed:
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE with AudioBufferList
https://bugs.webkit.org/show_bug.cgi?id=303743

Reviewed by Darin Adler.

* Source/WTF/wtf/StdLibExtras.h:
(WTF::lazyInitialize):
* Source/WTF/wtf/SystemFree.h:
(WTF::SystemFree::operator() const):
* Source/WTF/wtf/SystemMalloc.h:
(WTF::SystemMallocBase::zeroedMalloc):
* Source/WebCore/Modules/webaudio/MediaStreamAudioSourceCocoa.cpp:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/cf/CoreAudioExtras.h: Renamed from Source/WebCore/platform/audio/cocoa/SpanCoreAudio.h.
(PAL::allocationSize):
(PAL::createAudioBufferList):
* Source/WebCore/PAL/pal/spi/cf/CoreAudioSPI.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.cpp:
* Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.cpp:
(WebCore::AudioFileReader::decodeWebMData const):
(WebCore::AudioFileReader::createBus):
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.mm:
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.cpp:
* Source/WebCore/platform/audio/cocoa/CARingBuffer.cpp:
* Source/WebCore/platform/audio/cocoa/WebAudioBufferList.cpp:
(WebCore::WebAudioBufferList::WebAudioBufferList):
(WebCore::WebAudioBufferList::reset):
* Source/WebCore/platform/audio/cocoa/WebAudioBufferList.h:
* Source/WebCore/platform/audio/mac/AudioHardwareListenerMac.cpp:
* Source/WebCore/platform/audio/mac/AudioSessionMac.mm:
* Source/WebCore/platform/graphics/avfoundation/AudioSourceProviderAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/AudioSourceProviderAVFObjC.mm:
(WebCore::AudioSourceProviderAVFObjC::prepare):
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererInternalUnit.cpp:
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureUnit.cpp:
* Source/WebCore/platform/mediastream/mac/MockAudioCaptureUnit.mm:
* Source/WebCore/platform/mediastream/mac/RealtimeOutgoingAudioSourceCocoa.cpp:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::renderAudio):

Canonical link: https://commits.webkit.org/304157@main
